node/core/av-store/src/lib.rs

```diff
@@ -53,374 +57,353 @@ mod columns {
+#[derive(Debug, Clone, PartialEq, PartialOrd, Eq, Ord)]
+struct BEBlockNumber(BlockNumber);
+
+impl Encode for BEBlockNumber {
+	fn size_hint(&self) -> usize {
+		std::mem::size_of::<BlockNumber>()
 	}
 
-	fn as_duration(&self) -> Option<Duration> {
-		match self {
-			PruningDelay::In(d) => Some(*d),
-			PruningDelay::Indefinite => None,
-		}
+	fn using_encoded<R, F: FnOnce(&[u8]) -> R>(&self, f: F) -> R {
+		f(&self.0.to_be_bytes())
 	}
 }
 
-impl From<Duration> for PruningDelay {
-	fn from(d: Duration) -> Self {
-		Self::In(d)
+impl Decode for BEBlockNumber {
+	fn decode<I: Input>(value: &mut I) -> Result<Self, CodecError> {
+		<[u8; std::mem::size_of::<BlockNumber>()]>::decode(value).map(BlockNumber::from_be_bytes).map(Self)
 	}
 }
 
-impl PartialEq for PruningDelay {
-	fn eq(&self, other: &Self) -> bool {
-		match (self, other) {
-			(PruningDelay::In(this), PruningDelay::In(that)) => {this == that},
-			(PruningDelay::Indefinite, PruningDelay::Indefinite) => true,
-			_ => false,
-		}
-	}
+#[derive(Debug, Encode, Decode)]
+enum State {
+	/// Candidate data was first observed at the given time but is not available in any block.
+	Unavailable(BETimestamp),
+	/// The candidate was first observed at the given time and was included in the given list of unfinalized blocks, which may be
+	/// empty. The timestamp here is not used for pruning. Either one of these blocks will be finalized or the state will regress to
+	/// `State::Unavailable`, in which case the same timestamp will be reused.
+	Unfinalized(BETimestamp, Vec<(BEBlockNumber, Hash)>),
+	/// Candidate data has appeared in a finalized block and did so at the given time.
+	Finalized(BETimestamp)
 }
 
-impl PartialOrd for PruningDelay {
-	fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
-		match (self, other) {
-			(PruningDelay::In(this), PruningDelay::In(that)) => this.partial_cmp(that),
-			(PruningDelay::In(_), PruningDelay::Indefinite) => Some(Ordering::Less),
-			(PruningDelay::Indefinite, PruningDelay::In(_)) => Some(Ordering::Greater),
-			(PruningDelay::Indefinite, PruningDelay::Indefinite) => Some(Ordering::Equal),
-		}
-	}
+// Meta information about a candidate.
+#[derive(Debug, Encode, Decode)]
+struct CandidateMeta {
+	state: State,
+	data_available: bool,
+	chunks_stored: BitVec<BitOrderLsb0, u8>,
 }
 
-impl Ord for PruningDelay {
-	fn cmp(&self, other: &Self) -> Ordering {
-		match (self, other) {
-			(PruningDelay::In(this), PruningDelay::In(that)) => this.cmp(that),
-			(PruningDelay::In(_), PruningDelay::Indefinite) => Ordering::Less,
-			(PruningDelay::Indefinite, PruningDelay::In(_)) => Ordering::Greater,
-			(PruningDelay::Indefinite, PruningDelay::Indefinite) => Ordering::Equal,
+fn query_inner<D: Decode>(
+	db: &Arc<dyn KeyValueDB>,
+	column: u32,
+	key: &[u8],
+) -> Option<D> {
+	match db.get(column, key) {
+		Ok(Some(raw)) => {
+			let res = D::decode(&mut &raw[..]).expect("all stored data serialized correctly; qed");
```

Comment:
I see that this comes from the previous implementation, but maybe we should revisit this? In other similar places we prefer not to panic